handler/src/main/java/io/netty/handler/ssl/OpenSslKeyMaterialManager.java

```diff
@@ -0,0 +1,121 @@
+    private static final Map<String, String> KEY_TYPES = new HashMap<String, String>();
+    static {
+        KEY_TYPES.put("RSA", KEY_TYPE_RSA);
+        KEY_TYPES.put("DHE_RSA", KEY_TYPE_RSA);
+        KEY_TYPES.put("ECDHE_RSA", KEY_TYPE_RSA);
+        KEY_TYPES.put("ECDHE_ECDSA", KEY_TYPE_EC);
+        KEY_TYPES.put("ECDH_RSA", KEY_TYPE_EC_RSA);
+        KEY_TYPES.put("ECDH_ECDSA", KEY_TYPE_EC_EC);
+        KEY_TYPES.put("DH_RSA", KEY_TYPE_DH_RSA);
+    }
+
+    private final X509KeyManager keyManager;
+    private final String password;
+
+    OpenSslKeyMaterialManager(X509KeyManager keyManager, String password) {
+        this.keyManager = keyManager;
+        this.password = password;
+    }
+
+    void setKeyMaterial(OpenSslEngine engine) throws SSLException {
+        long ssl = engine.sslPointer();
+        String[] authMethods = SSL.authenticationMethods(ssl);
+        for (String authMethod : authMethods) {
+            String type = KEY_TYPES.get(authMethod);
+            if (type != null) {
+                setKeyMaterial(ssl, chooseServerAlias(engine, type));
+            }
+        }
+    }
+
+    void setKeyMaterial(OpenSslEngine engine, String[] keyTypes, X500Principal[] issuer) throws SSLException {
+        setKeyMaterial(engine.sslPointer(), chooseClientAlias(engine, keyTypes, issuer));
+    }
+
+    private void setKeyMaterial(long ssl, String alias) throws SSLException {
+        long keyBio = 0;
+        long keyCertChainBio = 0;
+        try {
+            // TODO: Should we cache these and so not need to do a memory copy all the time ?
+            PrivateKey key = keyManager.getPrivateKey(alias);
+            X509Certificate[] certificates = keyManager.getCertificateChain(alias);
+
+            if (certificates != null && certificates.length != 0) {
+                keyCertChainBio = OpenSslContext.toBIO(keyManager.getCertificateChain(alias));
+                if (key != null) {
+                    keyBio = OpenSslContext.toBIO(key);
+                }
+                SSL.setCertificateBio(ssl, keyCertChainBio, keyBio, password);
+
+                // We may have more then one cert in the chain so add all of them now.
+                SSL.setCertificateChainBio(ssl, keyCertChainBio, false);
+            }
+        } catch (SSLException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new SSLException(e);
+        } finally {
+            if (keyBio != 0) {
+                SSL.freeBIO(keyBio);
+            }
+            if (keyCertChainBio != 0) {
+                SSL.freeBIO(keyCertChainBio);
+            }
+        }
+    }
+
+    protected String chooseClientAlias(@SuppressWarnings("unused") OpenSslEngine engine,
+                                       String[] keyTypes, X500Principal[] issuer) {
+        return keyManager.chooseClientAlias(keyTypes, issuer, null);
+    }
+
+    protected String chooseServerAlias(@SuppressWarnings("unused") OpenSslEngine engine, String type) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "engine". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)